iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/builtin/relational/tvf/CumulateTableFunction.java

```diff
@@ -0,0 +1,133 @@
+            .build());
+  }
+
+  @Override
+  public TableFunctionAnalysis analyze(Map<String, Argument> arguments) throws UDFException {
+    // size must be an integral multiple of step.
+    long size = (long) ((ScalarArgument) arguments.get(SIZE_PARAMETER_NAME)).getValue();
+    long step = (long) ((ScalarArgument) arguments.get(STEP_PARAMETER_NAME)).getValue();
+
+    if (size % step != 0) {
+      throw new UDFException(
+          "Cumulative table function requires size must be an integral multiple of step.");
+    }
+
+    TableArgument tableArgument = (TableArgument) arguments.get(DATA_PARAMETER_NAME);
+    String expectedFieldName =
+        (String) ((ScalarArgument) arguments.get(TIMECOL_PARAMETER_NAME)).getValue();
+    int requiredIndex =
+        findColumnIndex(tableArgument, expectedFieldName, Collections.singleton(Type.TIMESTAMP));
+    DescribedSchema properColumnSchema =
+        new DescribedSchema.Builder()
+            .addField("window_start", Type.TIMESTAMP)
+            .addField("window_end", Type.TIMESTAMP)
+            .build();
+
+    // outputColumnSchema
+    return TableFunctionAnalysis.builder()
+        .properColumnSchema(properColumnSchema)
+        .requireRecordSnapshot(false)
+        .requiredColumns(DATA_PARAMETER_NAME, Collections.singletonList(requiredIndex))
+        .build();
+  }
+
+  @Override
+  public TableFunctionProcessorProvider getProcessorProvider(Map<String, Argument> arguments) {
+    return new TableFunctionProcessorProvider() {
+      @Override
+      public TableFunctionDataProcessor getDataProcessor() {
+        return new CumulateDataProcessor(
+            (Long) ((ScalarArgument) arguments.get(ORIGIN_PARAMETER_NAME)).getValue(),
+            (Long) ((ScalarArgument) arguments.get(STEP_PARAMETER_NAME)).getValue(),
+            (Long) ((ScalarArgument) arguments.get(SIZE_PARAMETER_NAME)).getValue());
+      }
+    };
+  }
+
+  private static class CumulateDataProcessor implements TableFunctionDataProcessor {
+
+    private final long step;
+    private final long size;
+    private final long start;
+    private long curIndex = 0;
+
+    public CumulateDataProcessor(long startTime, long step, long size) {
+      this.step = step;
+      this.size = size;
+      this.start = startTime;
+    }
+
+    @Override
+    public void process(
+        Record input,
+        List<ColumnBuilder> properColumnBuilders,
+        ColumnBuilder passThroughIndexBuilder) {
+      // find the first windows
+      long timeValue = input.getLong(0);
+      long window_start = (timeValue - start) / size * size;
+      long steps = step;
+      while (window_start <= timeValue && window_start + steps <= window_start + size) {
+        if (window_start + steps > timeValue) {
+          properColumnBuilders.get(0).writeLong(window_start);
+          properColumnBuilders.get(1).writeLong(window_start + steps);
+          passThroughIndexBuilder.writeLong(curIndex);
+        }
+        steps += step;
+      }
```

Comment:
```suggestion
      for(long steps = (timeValue-window_start+step)/step*step; steps<=size; steps+=step){
          properColumnBuilders.get(0).writeLong(window_start);
          properColumnBuilders.get(1).writeLong(window_start + steps);
          passThroughIndexBuilder.writeLong(curIndex);
      }
```

Maybe it can be implemented by simpler loop?

Reply:
Right